File: intern/1249.minimum-remove-to-make-valid-parentheses.js
/*
 * @lc app=leetcode id=1249 lang=javascript
 *
 * [1249] Minimum Remove to Make Valid Parentheses
 */

// @lc code=start
/**
 * @param {string} s
 * @return {string}
 */
var minRemoveToMakeValid = function (s) {
    const chars = [...s];
    /**
     * @type {['('|')',number][]} - [括号，下标]
     */
    const stack = [];

    for (let i = 0; i < chars.length; i++) {
        if (chars[i] === '(' || chars[i] === ')') {
            if (chars[i] === '(') {
                stack.push(['(', i]);
            } else if (chars[i] === ')') {
                if (stack.length === 0 || stack[stack.length - 1][0] === ')') {
                    stack.push([')', i]);
                } else {
                    stack.pop();
                }
            }
        }
    }

    for (const [, index] of stack) {
        chars[index] = '';
    }

    return chars.join('');
};
// @lc code=end

File: Daily/1249.minimum-remove-to-make-valid-parentheses.cpp
/*
 * @lc app=leetcode id=1249 lang=cpp
 *
 * [1249] Minimum Remove to Make Valid Parentheses
 */
#include <stack>
#include <string>
#include <unordered_set>

using std::stack;
using std::string;
using std::unordered_set;

// @lc code=start
class Solution {
 public:
  string minRemoveToMakeValid(string s) {
    stack<int> leftParenthesesIndexes;

    for (int i = 0; i < s.size(); i++) {
      if (s[i] == '(') {
        leftParenthesesIndexes.push(i);
      } else if (s[i] == ')') {
        if (!leftParenthesesIndexes.empty()) {
          leftParenthesesIndexes.pop();
        } else {
          s[i] = '*';
        }
      }
    }

    while (!leftParenthesesIndexes.empty()) {
      s[leftParenthesesIndexes.top()] = '*';
      leftParenthesesIndexes.pop();
    }

    string validString;
    for (char& c : s) {
      if (c != '*') {
        validString.push_back(c);
      }
    }

    return validString;
  }
};
// @lc code=end
